torch_geometric/nn/conv/peg_conv.py

```diff
@@ -0,0 +1,196 @@
+        else:
+            self.register_parameter('bias', None)
+
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        self.glorot(self.weight_withformer)
+        self.glorot(self.weight_noformer)
+        self.zeros(self.bias)
+        self._cached_edge_index = None
+        self._cached_adj_t = None
+
+    def forward(self, x: Tensor, edge_index: Adj,
+                edge_weight: OptTensor = None) -> Tensor:
+        """"""
+
+        coors, feats = x[:, :self.pos_dim], x[:, self.pos_dim:]
+
+        if self.normalize:
+            if isinstance(edge_index, Tensor):
+                cache = self._cached_edge_index
+                if cache is None:
+                    edge_index, edge_weight = gcn_norm(  # yapf: disable
+                        edge_index, edge_weight, feats.size(self.node_dim),
+                        self.improved, self.add_self_loops)
+                    if self.cached:
+                        self._cached_edge_index = (edge_index, edge_weight)
+                else:
+                    edge_index, edge_weight = cache[0], cache[1]
+
+            elif isinstance(edge_index, SparseTensor):
+                cache = self._cached_adj_t
+                if cache is None:
+                    edge_index = gcn_norm(  # yapf: disable
+                        edge_index, edge_weight, feats.size(self.node_dim),
+                        self.improved, self.add_self_loops)
+                    if self.cached:
+                        self._cached_adj_t = edge_index
+                else:
+                    edge_index = cache
+        else:
+            print('We normalize the adjacency matrix in PEG.')
+
+        if isinstance(edge_index, Tensor):
+            rel_coors = coors[edge_index[0]] - coors[edge_index[1]]
+        elif isinstance(edge_index, SparseTensor):
+            rel_coors = coors[
+                edge_index.to_torch_sparse_coo_tensor()._indices()[0]] - coors[
+                    edge_index.to_torch_sparse_coo_tensor()._indices()[1]]
+        rel_dist = (rel_coors**2).sum(dim=-1, keepdim=True)
+
+        # propagate_type: (x: Tensor, edge_weight: OptTensor)
+        # pos: l2 norms
+        hidden_out, coors_out = self.propagate(edge_index, x=feats,
+                                               edge_weight=edge_weight,
+                                               pos=rel_dist, coors=coors,
+                                               size=None)
+
+        if self.bias is not None:
+            hidden_out += self.bias
+
+        return torch.cat([coors_out, hidden_out], dim=-1)
+
+    def message(self, x_i: Tensor, x_j: Tensor, edge_weight: OptTensor,
+                pos) -> Tensor:
+        PE_edge_weight = self.edge_mlp(pos)
+        return x_j if edge_weight is None else (PE_edge_weight *
+                                                edge_weight.view(-1, 1) * x_j)
+
+    def propagate(self, edge_index: Adj, size: Size = None, **kwargs):
```

Comment:
Sorry I didn't fully follow the implementation yet - but I think you should be able to do this without needing to reimplement `propagate` - I.e you can put most of this logic inside your forward function.

Reply:
+1 . We definitely do not want to override `propagate` here. Any reason you did this in the first place?

Reply:
Thanks for your advice. I put the operational logic inside my forward function.